### ISSUE TITLE
Unquoted color names are confusing

### ISSUE THREAD

**Issue by hqx** [sass/sass#1256]
with style: compressed, the replacement of some colornames to hex codes takes place even in map keys. This is unexpected if not a bug.

Example below compiles uncompressed, but gets a syntax error compressed.

Syntax error: Invalid CSS after "&.": expected class name, was "#90ee90"
        on line 14 of demo.scss

(version 3.3.7)

```
$schemes: (
        lightgreen: (
                primary: rgb(30, 151, 64),
                secondary: rgb(149, 0, 70)
        )
);

a {
  text-decoration: none;
  color:  #999;
  @each $name, $spec in $schemes {
    &.#{$name} { color: map-get($spec, primary) }
  }
}

```

**Comment by nex3** [comment]
Color names refer to color objects in SassScript. If you want to use strings as map keys, you should quote them.

**Comment by hqx** [comment]
as I said, the bahavoiur is not expected nor consistent. 
It compiles fine when not compressed. 
And a "color object" blue works well, even with compression. 
So, that "lightgreen" suddenly is a problem, is not easily expected. 
An I have not yet seen docu or examples, wher the map keys are put in quotes. So it's not clear that that is even valid syntax. Let alone needed, in 4% random cases ...

**Comment by nex3** [comment]
It's entirely consistent. In compressed mode, Sass will choose the most compact representation when emitting a color. For `blue`, this is `blue`; for `lightgreen`, this is `#90ee90`.

**Comment by honzie** [comment]
+1 to this feels inconsistent. I spent time on the site today looking for documentation on this, and found this issue before cutting my own for a similar issue.

The documentation on maps is astoundingly light, and the code itself (https://github.com/nex3/sass/blob/dedd5b8ef35b36162ce5eda8319c04c7ea548659/lib/sass/script/functions.rb#L1916) doesn't mention the ability for things to not be input without quotes, nor does it mention that HTML color names (of which there are 140) may not be interpreted as expected.

Are there other sets of key names that should be avoided using? Can this be documented on sass-lang.com and/or in the source code?

**Comment by honzie** [comment]
@nex3, you had comments here (https://github.com/nex3/sass/issues/1122#issuecomment-37700189) that succinctly describe why changing how key lookup works might be a good idea. Particularly: "If [value] doesn't mean the exact same thing in all contexts, even if the differences seem benign, it encourages users to believe that other values may mean different things in other contexts." and "...they may (reasonably) suspect that it's because of some other hidden behavior that they're unaware of. Once users start believing this, their belief that they understand what Sass is doing goes down, which means effectively that their understanding goes down as well. This is bad."

**Comment by chriseppstein** [comment]
A value is the same value irrespective of the context. if you write blue in one place in a sass file it means the same thing everywhere. We're not going to magically make blue not a color for map keys. That is a kind of consistency. I get that you didn't expect this to work this way, but the consistency you expect may not be the consistency someone else expects. Nathan's quote is a good one it's something we strive for. It's also something that we realize is not 100% achievable for every single person at first glance.

**Comment by hqx** [comment]
Well it's not the same, that's the problem. 
A class .lightgreen is perfectly valid, but with compress, sass destroys that and then complains about syntax error. It's just a search- and replace action that works a bit too carelessly, I would say.

**Comment by chriseppstein** [comment]
You're saying that you think

``` scss
$c: lightgreen;
.#{$c} {
  color: #{$c};
}
```

Should produce in compressed mode:

``` css
.lightgreen{color:#90ee90;}
```

Am I correct?

**Comment by hqx** [comment]
I'm not sure if I'm a big fan of the name-to-hex-replacement-magic at all. ;-)
But that aside, yes, it should work like you say: replace colors only where it's really a color, not a class name or something else.  
I can see: with variables and maps this gets more complicated. 
I think the replacement should be done last after all variable expansions, not before, as it seems to be now.
I do get your point with the variable holding a color (maybe, still not 100% consistent, when unquoted foobar is a string, why must unquoted cornsilk absolutely be a color; who remembers all those silly names?)
But let's say I get your point with variables. A map _key_ still is something different. In not many languages a key like that can be an object, I would expect that to be a label/string. 
But maybe it's just me ...

Now I know, I have no problem to quote my keys. It should be better documented though.

**Comment by Arcovion** [comment]
> A map key still is something different. In not many languages a key like that can be an object, I would expect that to be a label/string. 

I agree with this, if it were a value like the `rgb()` you have then it'd be different - but allowing colour objects as keys is strange. It's like allowing the `rgb()` function as a key, does that work too? =]

I would expect both to raise an exception.

Edit: Also,

> Example below compiles uncompressed, but gets a syntax error compressed.

This should never happen - syntax errors shouldn't occur just because you switch compression modes.

**Comment by chriseppstein** [comment]
> It's like allowing the rgb() function as a key, does that work too? =]
> Yes. you can use rgb() as a key. it returns a color. `map-get($m, rgb(0, 128, 0))`, `map-get($m, green)`, and `map-get($m, #008000)` all return the same value stored against the same color value in the map.
> 
> I would expect both to raise an exception.

I would expect it to work. A map is an association between values. Limiting keys to strings is very inconvenient for a lot of use cases.

> This should never happen - syntax errors shouldn't occur just because you switch compression modes.

This happens because #{} says evaluate this expression and put it's string value right here. The string value for colors varies by compression mode. This happens because you put a color in a selector but selectors only allow strings and combinators.

This isn't a problem with maps. This is a problem that you're expecting sass to treat named colors as sometimes a string and sometimes a color. We decided a long time ago that we're not going to do that. Colors are always colors, sass is playing dumb here because dumb things, while sometimes cumbersome, are easy to reason about.

TL;DR put quotes around your named colors if you want them to be strings.

I'm not sure that we're going to convince you on this matter. We have heard your arguments. They are good ones, but we think the way it works right now is, on the balance, better than what you're proposing.

**Comment by Arcovion** [comment]
I was under the impression that map keys _were_ for strings, as maps allow strings without needing quotes - but I forgot this is Sass, which allows anything to be a string! That makes this very difficult to address. 
You're right that this isn't a problem with maps, but I think it _is_ a problem with compression:

Sass is very clever with it's optimization, [I learned recently](https://github.com/nex3/sass/issues/1253#issuecomment-43558436) that 

> Sass will only include a @charset declaration in the CSS output if there are any non-ASCII characters.

which is cool and all, but it's another example of unexpected behaviour.

Things like this make me want to use a normal CSS minifier instead, but I suppose that's the point of `:compressed` mode - to perform aggressive optimisations that normal minifiers can't do...

So, what about adding another output option? One that doesn't convert colour names to hex, and leaves the charset in just for the [principle of least surprise](http://en.wikipedia.org/wiki/Principle_of_least_astonishment); like a normal minifier would, in fact designed to act exactly like a normal minifier.

**Comment by chriseppstein** [comment]
I have [clarified this behavior in the Sass Reference](https://github.com/nex3/sass/compare/a2fd37e...a76cb3f).

**Comment by hqx** [comment]
> So, what about adding another output option? One that doesn't convert colour names to hex, and leaves ...

I agree.

My first surprise was that compression would change the content at all. More so if this introduces problems that were not there before.

If you are serious about 
`$c: lightgreen;` 
being a color object, not a string, then 
`.#{$c}` 
should be a syntax error no matter the output style.
That output syle decides wether the input file has a valid syntax or not, is unprecedented. I think.

**Comment by chriseppstein** [comment]
@nex3 I'm having a hard time justifying saying no here given that we automatically unquote strings when being interpolated in a selector. An error in all cases would be a better experience, IMO.

**Comment by nex3** [comment]
First of all, it's clear that supporting unquoted map keys causes a substantial amount of confusion with colors. I'm very tempted to just make Sass emit a warning whenever _either_ an unquoted string or a color is used as a map key in a map literal. This does violate the invariant that any Sass value can be used as a key, but I think (given that these keys may be added programmatically) the benefit of driving users towards quoted keys will be worth the pain.

@chriseppstein, I don't understand the relevance of automatically unquoting strings.

**Comment by chriseppstein** [comment]
> Sass emit a warning whenever either an unquoted string or a color is used as a map key in a map literal.

:-1: That is going to cause a lot of heartache and ugly code.

@nex3 this isn't a map problem. it's a color output problem. that has been the source confusion for quite some time even before maps. The issue only matters when the keys are placed into selectors.

**Comment by nex3** [comment]
> :-1: That is going to cause a lot of heartache and ugly code.

By "ugly code" do you mean quoted map keys? These are used in many languages without issue.

> this isn't a map problem. it's a color output problem. that has been the source confusion for quite some time even before maps. The issue only matters when the keys are placed into selectors.

I agree that maps are tangential to the core issue, although I'd say that issue is that named color literals and unquoted string literals are visually indistinguishable. Maps bring this issue to the foreground, because they (and we) encourage users to use unquoted strings for the keys, and those keys make their way into selectors much more frequently than color values ever did before. But the underlying issue is this confusion, and selector interpolation is merely the most common way it manifests.

Unfortunately, we're locked into supporting both unquoted string literals and named color literals by the CSS syntax, so we can't address the issue head-on. Instead, we need to find ways to mitigate it, in some combination of encouraging users to use unambiguous notation and punishing them less when they do.

5bba828be89461043564911a7cdf08eec35dc359 addresses the latter to some extent, ensuring that outside of compressed mode named colors are consistently emitted in their named form. My proposal addresses the former, by providing messaging to users that unquoted strings are likely to behave in unpredictable ways. That messaging is tied maps because they're empirically one of the most likely places to actually produce the errors in question.

I admit, though, that I don't completely understand your proposed behavior, so perhaps it addresses these issues too.

**Comment by chriseppstein** [comment]
@nex3 We currently automatically unquote any quoted string that is interpolated into a selector because we know that quotes are not legal in that context. So my point is that we already are special-casing output of interpolation in a selector.

I see two possible approaches. We make it so that in a selector context named colors always output as a string instead of as a color irrespective of the output format if the color is a named color. If the color is not a named color, it would be illegal syntax in all output modes. Alternatively, we simply outlaw color being output in a selector context even for named colors.

**Comment by nex3** [comment]
> We currently automatically unquote any quoted string that is interpolated into a selector because we know that quotes are not legal in that context. So my point is that we already are special-casing output of interpolation in a selector.

This isn't quite accurate. We unquote any quoted string that's interpolated _anywhere_. It's part of the semantics of interpolation, not a special case for selectors in particular.

> I see two possible approaches. We make it so that in a selector context named colors always output as a string instead of as a color irrespective of the output format if the color is a named color. If the color is not a named color, it would be illegal syntax in all output modes. Alternatively, we simply outlaw color being output in a selector context even for named colors.

I'm uncomfortable having selector-specific stringification semantics or errors. I'm more comfortable having context-sensitive warnings, though; I could see warning if a color was passed directly to interpolation within a selector.

**Comment by lolmaus** [comment]
People keep running into this.

As it has been noted above, the issue is not with map keys, but with putting colors into selectors.

Why don't you have Sass spit out a warning whenever a color is being interpolated inside a selector? This would resolve the issue for everybody and it does not require abandoning the ability of using any object as a map key.

**Comment by nex3** [comment]
> Why don't you have Sass spit out a warning whenever a color is being interpolated inside a selector? This would resolve the issue for everybody and it does not require abandoning the ability of using any object as a map key.

In fact that's what I suggested above :smiley:.

**Comment by shadowmint** [comment]
This should be a warning.

Is there any case where you would _actually_ want your map keys to be colour values instead of meaningful (ie. string) keys?

I think it's pretty obvious from the extent of this thread and that it keep coming up, that simply updating the documentation isn't really a great solution.

**Comment by cimmanon** [comment]
> simply updating the documentation isn't really a great solution.

People read the documentation?

**Comment by barneycarroll** [comment]
This had me confused for ages. The fact that Sass maps can take any reference as keys makes sense of the notion that unquoted keys are interpolated. What is very surprising is that Sass takes it upon itself to parse color name references — which, in the intended scenario, is simply raw CSS — and compile them to hex values. To my mind this is the same kind of design mistake as hijacking the `@import` directive: a user writing plain valid CSS should not expect Sass to second-guess their intent and provide its own interpretation. The byte-shaving this might provide in certain circumstances is trivial, not at all worth this weird functional gotcha: the obfuscation of production code is needless, and moreover Sass has no business deciding how CSS colours are mapped & interpreted. At best, this 'feature' means Sass needs to commit to otherwise redundant updates in step with the CSS Colors module specification.

The difference here is that fixing this mistake won't negatively impact anybody, beyond perhaps adding several bytes to some people's compiled code.

Logging a warning is a good first step, but this should be on the road to deprecation. AFAICT no users have expressed anything other than surprise & frustration at this.

**Comment by nex3** [comment]
Sass interprets property values in the same way CSS does. CSS uses color names to refer to colors, so Sass does as well. I guarantee that if it didn't, we'd have a lot more people wondering why color functions don't work with named colors than we have people upset about color interpolation now.

**Comment by barneycarroll** [comment]
@nex3 good point!

> Sass interprets property values in the same way CSS does

This is a crucial insight. The problem case is when Sass interprets things which aren't property values (they're keys in a map, which is a data structure with no CSS equivalent) as if they were.

> why color functions don't work with named colors

This is perhaps they key to fixing this. The only time Sass functionality depends upon colour interpolation is when is when it needs to operate upon said colours. This is similar to how jQuery methods involving animation will try to interpret a passed in strings of `slow` & `fast` as indicating a duration of 600 or 200 milliseconds respectively.

These insights suggests 2 concrete methods for fixing the issue:
1. Make colour string interpretation a feature of colour functions - ie only perform it when we need to manipulate colours
2. Only interpret colour strings when they are used as property values

There is still a problem space around the fact that 'colors' are a basic type in Sass, which raises some expectations. The documentation makes explicit reference to this (and the issue that you have to quote CSS-valid color names if you intend to use them as strings) [here](http://sass-lang.com/documentation/file.SASS_REFERENCE.html#colors).

The reason this is problematic is that users might write their own functionality which depends upon `type-of( red ) === color` and pre-conversion for manipulation. But AFAICT this does not predicate that all string color references be pre-converted to hex references - it just demands that the color type be extended to allow and recognise strings. So the solution then seems to be:
- Recognise CSS color string input as entities of type color, but do not prematurely transform them into shortest possible hex representation by default. 

We already have things like alpha manipulation on strings which may otherwise be output as short-form hex, so it's not as if Sass isn't capable of intelligently keeping color references in memory as flexible types capable of a variety of outputs. So I would prefer us to stick to the first solution, and fall back to the second if that isn't possible - in which case we'd probably have to introduce coercion logic for string comparison (such that `red == 'red'`, similar to how `1 == '1'`).

I tried reading the test suite to see if these expectations are tested for, but couldn't really make heads or tails of it. Would a straight functional PR be good, or do we need to write the tests that guard against violating these expectations?

**Comment by barneycarroll** [comment]
@nex3 sorry, just reading back through the thread again - is this suggestion what the v4 alpha commit [here](https://github.com/sass/sass/commit/5bba828be89461043564911a7cdf08eec35dc359) is about?

**Comment by nex3** [comment]
I'm not interested in changing the existing behavior. It does have some awkward edge-cases, but the underlying logic is clear and unambiguous—colors and strings are wholly separate types, and the implementation has license to treat _all_ colors the same regardless of their literal format.

> `1 == '1'`

This returns `false`. Sass does no type coercion in `==`.

> the v4 alpha commit [here](https://github.com/sass/sass/commit/5bba828be89461043564911a7cdf08eec35dc359) is about?

No. That commit is purely cosmetic; it makes colors more similar to their source definition in human-readable output formats, but it doesn't change the language itself. In particular, in compressed mode, Sass will still choose the smallest possible format for a color. This may be behavior users are relying on, and I'm not willing to change it.
